lib/util/usedPropTypes.js

```diff
@@ -0,0 +1,526 @@
+// Constants
+// ------------------------------------------------------------------------------
+
+const DIRECT_PROPS_REGEX = /^props\s*(\.|\[)/;
+const DIRECT_NEXT_PROPS_REGEX = /^nextProps\s*(\.|\[)/;
+const DIRECT_PREV_PROPS_REGEX = /^prevProps\s*(\.|\[)/;
+const LIFE_CYCLE_METHODS = ['componentWillReceiveProps', 'shouldComponentUpdate', 'componentWillUpdate', 'componentDidUpdate'];
+const ASYNC_SAFE_LIFE_CYCLE_METHODS = ['getDerivedStateFromProps', 'getSnapshotBeforeUpdate', 'UNSAFE_componentWillReceiveProps', 'UNSAFE_componentWillUpdate'];
+
+/**
+ * Checks if a prop init name matches common naming patterns
+ * @param {ASTNode} node The AST node being checked.
+ * @returns {Boolean} True if the prop name matches
+ */
+function isPropAttributeName (node) {
+  return (
+    node.init.name === 'props' ||
+    node.init.name === 'nextProps' ||
+    node.init.name === 'prevProps'
+  );
+}
+
+/**
+ * Checks if the component must be validated
+ * @param {Object} component The component to process
+ * @returns {Boolean} True if the component must be validated, false if not.
+ */
+function mustBeValidated(component) {
+  return Boolean(
+    component &&
+    !component.ignorePropsValidation
+  );
+}
+
+module.exports = function usedPropTypesInstructions(context, components, utils) {
+  const sourceCode = context.getSourceCode();
+  const checkAsyncSafeLifeCycles = versionUtil.testReactVersion(context, '16.3.0');
+
+  /**
+   * Check if we are in a class constructor
+   * @return {boolean} true if we are in a class constructor, false if not
+   */
+  function inComponentWillReceiveProps() {
+    let scope = context.getScope();
+    while (scope) {
+      if (
+        scope.block && scope.block.parent &&
+        scope.block.parent.key && scope.block.parent.key.name === 'componentWillReceiveProps'
+      ) {
+        return true;
+      }
+      scope = scope.upper;
+    }
+    return false;
+  }
+
+  /**
+   * Check if we are in a lifecycle method
+   * @return {boolean} true if we are in a class constructor, false if not
+   **/
+  function inLifeCycleMethod() {
+    let scope = context.getScope();
+    while (scope) {
+      if (scope.block && scope.block.parent && scope.block.parent.key) {
+        const name = scope.block.parent.key.name;
+
+        if (LIFE_CYCLE_METHODS.indexOf(name) >= 0) {
+          return true;
+        } else if (checkAsyncSafeLifeCycles && ASYNC_SAFE_LIFE_CYCLE_METHODS.indexOf(name) >= 0) {
+          return true;
+        }
+      }
+      scope = scope.upper;
+    }
+    return false;
+  }
+
+  /**
+   * Returns true if the given node is a React Component lifecycle method
+   * @param {ASTNode} node The AST node being checked.
+   * @return {Boolean} True if the node is a lifecycle method
+   */
+  function isNodeALifeCycleMethod(node) {
+    const nodeKeyName = (node.key || {}).name;
+
+    if (node.kind === 'constructor') {
+      return true;
+    } else if (LIFE_CYCLE_METHODS.indexOf(nodeKeyName) >= 0) {
```

Comment:
also here (ie, the `no-else-return` eslint rule)

Reply:
yeah, I wonder why this isn't warning